python/nav/alertengine/base.py

```diff
@@ -249,177 +258,243 @@ def _check_permissions(account, alert, alertsubscription, dupemap, _logger,
+    :type queued_alerts: List[AccountAlertQueue]
+    :param now: Set only when testing, to force what value should be considered the
+                current time.
+    :type now: datetime
+
+    :return: A tuple of stats on how the queued alerts were processed:
+             (accounts_sent_daily, accounts_sent_weekly, num_sent_alerts,
+              num_failed_sends, num_resolved_alerts_ignored)
+    """
     _logger = logging.getLogger('nav.alertengine.handle_queued_alerts')
 
     if not now:
         now = datetime.now()
 
-    # We want to keep track of wether or not any weekly or daily messages have
-    # been sent so that we can update the state of the users
-    # last_sent_daily/weekly
-    sent_weekly = []
-    sent_daily = []
+    # We want to keep track of which users have received their weekly or daily
+    # notifications during this queue run, so that their profile states can be updated
+    accounts_sent_weekly = set()
+    accounts_sent_daily = set()
 
     num_sent_alerts = 0
     num_resolved_alerts_ignored = 0
     num_failed_sends = 0
 
-    for queued_alert in queued_alerts:
-        send, daily, weekly = False, False, False
+    for queued_alert in queued_alerts:  # type: AccountAlertQueue
+        result = process_single_queued_notification(queued_alert, now=now)
 
-        try:
-            subscription = queued_alert.subscription
-        except AlertSubscription.DoesNotExist:
-            _logger.error('account queued alert %d does not have subscription, '
-                          'probably a legacy table row', queued_alert.id)
-            continue
+        # The result value is processed purely for statistics and record keeping
+        if result == DISPATCH_FAILED:
+            num_failed_sends += 1
+        elif result == DISPATCH_IGNORED:
+            num_resolved_alerts_ignored += 1
+        elif result in (DISPATCHED_NOW, DISPATCHED_WEEKLY, DISPATCHED_DAILY):
+            num_sent_alerts += 1
 
-        _logger.debug('Stored alert %d: Checking %s %s subscription %d',
-                      queued_alert.alert_id, queued_alert.account,
-                      subscription.get_type_display(), subscription.id)
+        if result == DISPATCHED_DAILY:
+            accounts_sent_daily.add(queued_alert.account)
+        elif result == DISPATCHED_WEEKLY:
+            accounts_sent_weekly.add(queued_alert.account)
 
-        try:
-            subscription.time_period.profile.alertpreference
-        except AlertPreference.DoesNotExist:
-            subscription = None
+        del queued_alert
+
+    del queued_alerts
+
+    return (accounts_sent_daily, accounts_sent_weekly, num_sent_alerts,
+            num_failed_sends, num_resolved_alerts_ignored)
 
-        if subscription is None:
-            _logger.info('Sending alert %d right away as the users profile has '
-                         'been disabled', queued_alert.alert_id)
+
+def process_single_queued_notification(queued_alert, now):
```

Comment:
This refactor makes me wish for a different diff algorithm. Bah. Not obvious that this function is mostly just split from the one above.

Reply:
I'm revieweing commit by commit now...

Reply:
Good. I suggested reviewing commit by commit in an out-of-band conversation, but it would probably have been prudent to mention it here also.

I think `git diff` can be made to make more readable diffs by detecting moved code here, but I'm not sure the same applies to the GitHub UI.